src/geometry_base/kd_tree_strategy.cc

```diff
@@ -0,0 +1,143 @@
+                                                                         double radius) const {
+  const auto lane_ids = this->ClosestLanes(inertial_position.xyz(), radius);
+  std::vector<api::RoadPositionResult> road_positions;
+  for (const auto& current_lane : lane_ids) {
+    const auto lane = rg_->ById().GetLane(current_lane);
+    const auto lane_position = lane->ToLanePosition(inertial_position);
+    if (lane_position.distance <= radius) {
+      api::RoadPositionResult road_position{
+          {lane, lane_position.lane_position}, lane_position.nearest_position, lane_position.distance};
+      road_positions.push_back(road_position);
+    }
+  }
+  return road_positions;
+}
+
+void KDTreeStrategy::Init() {
+  const auto lanes = rg_->ById().GetLanes();
+
+  std::deque<maliput::geometry_base::MaliputPoint> points;
+  for (const auto& lane : lanes) {
+    const auto lane_length = lane.second->length();
+    for (double s = 0; s <= lane_length; s += 0.1) {
+      const auto lane_bounds = lane.second->lane_bounds(s);
+      for (double r = lane_bounds.min(); r <= lane_bounds.max(); r += 0.1) {
+        const auto inertial_pos = lane.second->ToInertialPosition({s, r, 0. /* h */}).xyz();
+        const maliput::geometry_base::MaliputPoint point{{inertial_pos.x(), inertial_pos.y(), inertial_pos.z()},
+                                                         lane.first};
+        points.push_back(point);
+      }
+    }
+  }
+  kd_tree_ = (std::make_unique<math::KDTree3D<MaliputPoint>>(std::move(points)));
+}
+
+maliput::api::LaneId KDTreeStrategy::do_closest_lane(const maliput::math::Vector3& point) const {
+  // Given that this method should actually obtain the lanes within a tolerance radius. Afterwards, the lane
+  // could be obtained using the point being closer to the centerline.
+  const api::InertialPosition& inertial_position{point.x(), point.y(), point.z()};
+  const auto maliput_point = kd_tree_->Nearest(point);
+  double radius = std::sqrt(pow(point.x() - maliput_point.x(), 2) + pow(point.y() - maliput_point.y(), 2) +
+                            pow(point.z() - maliput_point.z(), 2)) +
+                  0.02;
+  const auto lane_ids = do_closest_lanes(point, radius);
+
+  auto lane_id_result = maliput_point.lane_id();
+  auto lane_position_result = rg_->ById().GetLane(lane_id_result)->ToLanePosition(inertial_position);
+  double min_distance = lane_position_result.distance;
+  const static double KLineraTolerance{1e-12};  // [m]
+
+  for (const auto& current_lane : lane_ids) {
+    const auto current_lane_position = rg_->ById().GetLane(maliput_point.lane_id())->ToLanePosition(inertial_position);
+    if (current_lane_position.distance - min_distance > KLineraTolerance) {
+      min_distance = current_lane_position.distance;
+      lane_position_result = current_lane_position;
+      lane_id_result = current_lane;
+      continue;
```

Comment:
Question here, I went with the unchained if conditions and continue statement because the conditions depend on each other. So I first have to check if the new distance is smaller than the one I had before as `min_distance`. If both distances are the same (within the linear_tolerance error) then I have to check the r value of each position and select the one that is closest to 0. In other words we are talking about a nested if-else conditions with an additional nest layer depending on how many 'closest_point' definitions we have. 

IMO the nested if-else conditions give less readability than checking each condition independently and that is why I went with that structure. Let me know what you think @agalbachicar @francocipollone .

Reply:
My comment here doesn't make sense anymore due to the implementation of a general method `IsNewRoadPositionResultCloser()`